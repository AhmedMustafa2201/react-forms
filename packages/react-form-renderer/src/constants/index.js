export const components = {
  BUTTON: 'button',
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  WIZARD: 'wizard',
  SWITCH: 'switch-field',
  TEXTAREA: 'textarea-field',
  SELECT: 'select-field',
  PLAIN_TEXT: 'plain-text',
  INPUT_ADDON_GROUP: 'input-addon-group',
  INPUT_ADDON_BUTTON_GROUP: 'input-addon-button-group',
};
  /** Validator functions placeholder */
export const validators = {
  REQUIRED: 'required-validator',
  /**
     * min length of the input value
     */
  MIN_LENGTH: 'min-length-validator',
  /**
     * max length of the input value
     */
  MAX_LENGTH: 'max-length-validator',
  /**
   * exact length of input value
   */
  EXACT_LENGTH: 'exact-length-validator',
  /**
     * minimum count of fileds in some dynamic list of fields
     */
  MIN_ITEMS_VALIDATOR: 'min-items-validator',
  /**
     * Minimum value of number input
     */
  MIN_NUMBER_VALUE: 'min-number-value',
  /**
     * Maximum value of number inpuy
     */
  MAX_NUMBER_VALUE: 'max-number-value',
  /**
     * Regexp pattern validator
     */
  PATTERN_VALIDATOR: 'pattern-validator',
  /**
   * URL validator
   */
  URL: 'url-validator',
};

export const layoutComponents = {
  FORM_WRAPPER: 'FormWrapper',
  BUTTON: 'Button',
  BUTTON_GROUP: 'ButtonGroup',
  TITLE: 'Title',
  DESCRIPTION: 'Description',
};

/**
 * @enum {String}
 */
export const dataTypes = {
  INTEGER: 'integer',
  FLOAT: 'float',
  NUMBER: 'number',
  BOOLEAN: 'boolean',
  STRING: 'string',
};
