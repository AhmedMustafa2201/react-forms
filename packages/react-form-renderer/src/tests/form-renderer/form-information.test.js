import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import FormRenderer from '../../form-renderer';
import { layoutComponents } from '../../constants';
import RendererContext from '../../form-renderer/renderer-context';

describe('<FormControls />', () => {
  let initialProps;
  let ContextWrapper;
  let layoutMapper;
  let formFieldsMapper;

  const schema = {
    fields: [],
  };

  beforeEach(() => {
    formFieldsMapper = {};

    layoutMapper = {
      [layoutComponents.BUTTON]: ({ label, buttonType, ...props }) => <button { ...props }>{ label }</button>,
      [layoutComponents.BUTTON_GROUP]: ({ children, ...props }) => <div { ...props }>{ children }</div>,
      [layoutComponents.TITLE]: ({ children }) => <h1>{ children }</h1>,
      [layoutComponents.DESCRIPTION]: ({ children }) => <div>{ children }</div>,
      [layoutComponents.FORM_WRAPPER]: ({ children }) => <form>{ children }</form>,
    };

    ContextWrapper = ({ children, ...props }) => (
      <RendererContext.Provider
        value={{ layoutMapper, ...props }}
      >
        { children }
      </RendererContext.Provider>
    );
    initialProps = {
      onSubmit: jest.fn(),
      formFieldsMapper,
      layoutMapper,
    };
  });

  const TITLE = 'tiiiitle';

  it('should render with title', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormRenderer { ...initialProps } schema={{ ...schema, title: TITLE }}/>
      </ContextWrapper>
    );
    expect(wrapper.find('h1').text().includes(TITLE)).toEqual(true);
  });

  it('should render with label', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormRenderer { ...initialProps } schema={{ ...schema, label: TITLE }}/>
      </ContextWrapper>
    );
    expect(wrapper.find('h1').text().includes(TITLE)).toEqual(true);
  });

  it('should render with description', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormRenderer { ...initialProps } schema={{ ...schema, description: 'Description' }}/>
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render with title and description', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormRenderer { ...initialProps } schema={{ ...schema, title: 'Title', description: 'Description' }}/>
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render without title and description', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormRenderer { ...initialProps } schema={ schema }/>
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
