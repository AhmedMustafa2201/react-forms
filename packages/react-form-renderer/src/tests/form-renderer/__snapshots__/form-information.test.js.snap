// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormControls /> should render with description 1`] = `
<ContextWrapper>
  <FormRenderer
    buttonsLabels={Object {}}
    clearOnUnmount={false}
    disableSubmit={Array []}
    formFieldsMapper={Object {}}
    formType="pf3"
    initialValues={Object {}}
    layoutMapper={
      Object {
        "Button": [Function],
        "ButtonGroup": [Function],
        "Col": [Function],
        "Description": [Function],
        "FormGroup": [Function],
        "FormWrapper": [Function],
        "Title": [Function],
      }
    }
    onSubmit={[MockFunction]}
    resetAble={false}
    schema={
      Object {
        "description": "Description",
        "fields": Array [],
      }
    }
    schemaType="default"
    showFormControls={true}
    uiSchema={Object {}}
  >
    <ReactFinalForm
      decorators={
        Array [
          [Function],
        ]
      }
      initialValues={Object {}}
      mutators={
        Object {
          "concat": [Function],
          "insert": [Function],
          "move": [Function],
          "pop": [Function],
          "push": [Function],
          "remove": [Function],
          "removeBatch": [Function],
          "shift": [Function],
          "swap": [Function],
          "unshift": [Function],
          "update": [Function],
        }
      }
      onSubmit={[MockFunction]}
      render={[Function]}
      subscription={
        Object {
          "pristine": true,
          "submitting": true,
          "valid": true,
        }
      }
    >
      <Component>
        <form>
          <Component>
            <div>
              Description
            </div>
          </Component>
          <FormControls
            buttonOrder={
              Array [
                "submit",
                "reset",
                "cancel",
              ]
            }
            canReset={false}
            canSubmit={false}
            cancelLabel="Cancel"
            handleSubmit={[Function]}
            resetLabel="Reset"
            submitLabel="Submit"
          >
            <ReactFinalForm(FormSpy)>
              <FormSpy
                reactFinalForm={
                  Object {
                    "batch": [Function],
                    "blur": [Function],
                    "change": [Function],
                    "destroyOnUnregister": false,
                    "focus": [Function],
                    "getFieldState": [Function],
                    "getRegisteredFields": [Function],
                    "getState": [Function],
                    "initialize": [Function],
                    "isValidationPaused": [Function],
                    "mutators": Object {
                      "concat": [Function],
                      "insert": [Function],
                      "move": [Function],
                      "pop": [Function],
                      "push": [Function],
                      "remove": [Function],
                      "removeBatch": [Function],
                      "shift": [Function],
                      "swap": [Function],
                      "unshift": [Function],
                      "update": [Function],
                    },
                    "pauseValidation": [Function],
                    "registerField": [Function],
                    "reset": [Function],
                    "resetFieldState": [Function],
                    "resumeValidation": [Function],
                    "setConfig": [Function],
                    "submit": [Function],
                    "subscribe": [Function],
                  }
                }
              >
                <Component>
                  <div>
                    <Component
                      key="form-submit"
                      label="Submit"
                      onClick={[Function]}
                      type="button"
                      variant="primary"
                    >
                      <button
                        onClick={[Function]}
                        type="button"
                        variant="primary"
                      >
                        Submit
                      </button>
                    </Component>
                  </div>
                </Component>
              </FormSpy>
            </ReactFinalForm(FormSpy)>
          </FormControls>
        </form>
      </Component>
    </ReactFinalForm>
  </FormRenderer>
</ContextWrapper>
`;

exports[`<FormControls /> should render with title 1`] = `
<ContextWrapper>
  <FormRenderer
    buttonsLabels={Object {}}
    clearOnUnmount={false}
    disableSubmit={Array []}
    formFieldsMapper={Object {}}
    formType="pf3"
    initialValues={Object {}}
    layoutMapper={
      Object {
        "Button": [Function],
        "ButtonGroup": [Function],
        "Col": [Function],
        "Description": [Function],
        "FormGroup": [Function],
        "FormWrapper": [Function],
        "Title": [Function],
      }
    }
    onSubmit={[MockFunction]}
    resetAble={false}
    schema={
      Object {
        "fields": Array [],
        "title": "Title",
      }
    }
    schemaType="default"
    showFormControls={true}
    uiSchema={Object {}}
  >
    <ReactFinalForm
      decorators={
        Array [
          [Function],
        ]
      }
      initialValues={Object {}}
      mutators={
        Object {
          "concat": [Function],
          "insert": [Function],
          "move": [Function],
          "pop": [Function],
          "push": [Function],
          "remove": [Function],
          "removeBatch": [Function],
          "shift": [Function],
          "swap": [Function],
          "unshift": [Function],
          "update": [Function],
        }
      }
      onSubmit={[MockFunction]}
      render={[Function]}
      subscription={
        Object {
          "pristine": true,
          "submitting": true,
          "valid": true,
        }
      }
    >
      <Component>
        <form>
          <Component>
            <div>
              Title
            </div>
          </Component>
          <FormControls
            buttonOrder={
              Array [
                "submit",
                "reset",
                "cancel",
              ]
            }
            canReset={false}
            canSubmit={false}
            cancelLabel="Cancel"
            handleSubmit={[Function]}
            resetLabel="Reset"
            submitLabel="Submit"
          >
            <ReactFinalForm(FormSpy)>
              <FormSpy
                reactFinalForm={
                  Object {
                    "batch": [Function],
                    "blur": [Function],
                    "change": [Function],
                    "destroyOnUnregister": false,
                    "focus": [Function],
                    "getFieldState": [Function],
                    "getRegisteredFields": [Function],
                    "getState": [Function],
                    "initialize": [Function],
                    "isValidationPaused": [Function],
                    "mutators": Object {
                      "concat": [Function],
                      "insert": [Function],
                      "move": [Function],
                      "pop": [Function],
                      "push": [Function],
                      "remove": [Function],
                      "removeBatch": [Function],
                      "shift": [Function],
                      "swap": [Function],
                      "unshift": [Function],
                      "update": [Function],
                    },
                    "pauseValidation": [Function],
                    "registerField": [Function],
                    "reset": [Function],
                    "resetFieldState": [Function],
                    "resumeValidation": [Function],
                    "setConfig": [Function],
                    "submit": [Function],
                    "subscribe": [Function],
                  }
                }
              >
                <Component>
                  <div>
                    <Component
                      key="form-submit"
                      label="Submit"
                      onClick={[Function]}
                      type="button"
                      variant="primary"
                    >
                      <button
                        onClick={[Function]}
                        type="button"
                        variant="primary"
                      >
                        Submit
                      </button>
                    </Component>
                  </div>
                </Component>
              </FormSpy>
            </ReactFinalForm(FormSpy)>
          </FormControls>
        </form>
      </Component>
    </ReactFinalForm>
  </FormRenderer>
</ContextWrapper>
`;

exports[`<FormControls /> should render with title and description 1`] = `
<ContextWrapper>
  <FormRenderer
    buttonsLabels={Object {}}
    clearOnUnmount={false}
    disableSubmit={Array []}
    formFieldsMapper={Object {}}
    formType="pf3"
    initialValues={Object {}}
    layoutMapper={
      Object {
        "Button": [Function],
        "ButtonGroup": [Function],
        "Col": [Function],
        "Description": [Function],
        "FormGroup": [Function],
        "FormWrapper": [Function],
        "Title": [Function],
      }
    }
    onSubmit={[MockFunction]}
    resetAble={false}
    schema={
      Object {
        "description": "Description",
        "fields": Array [],
        "title": "Title",
      }
    }
    schemaType="default"
    showFormControls={true}
    uiSchema={Object {}}
  >
    <ReactFinalForm
      decorators={
        Array [
          [Function],
        ]
      }
      initialValues={Object {}}
      mutators={
        Object {
          "concat": [Function],
          "insert": [Function],
          "move": [Function],
          "pop": [Function],
          "push": [Function],
          "remove": [Function],
          "removeBatch": [Function],
          "shift": [Function],
          "swap": [Function],
          "unshift": [Function],
          "update": [Function],
        }
      }
      onSubmit={[MockFunction]}
      render={[Function]}
      subscription={
        Object {
          "pristine": true,
          "submitting": true,
          "valid": true,
        }
      }
    >
      <Component>
        <form>
          <Component>
            <div>
              Title
            </div>
          </Component>
          <Component>
            <div>
              Description
            </div>
          </Component>
          <FormControls
            buttonOrder={
              Array [
                "submit",
                "reset",
                "cancel",
              ]
            }
            canReset={false}
            canSubmit={false}
            cancelLabel="Cancel"
            handleSubmit={[Function]}
            resetLabel="Reset"
            submitLabel="Submit"
          >
            <ReactFinalForm(FormSpy)>
              <FormSpy
                reactFinalForm={
                  Object {
                    "batch": [Function],
                    "blur": [Function],
                    "change": [Function],
                    "destroyOnUnregister": false,
                    "focus": [Function],
                    "getFieldState": [Function],
                    "getRegisteredFields": [Function],
                    "getState": [Function],
                    "initialize": [Function],
                    "isValidationPaused": [Function],
                    "mutators": Object {
                      "concat": [Function],
                      "insert": [Function],
                      "move": [Function],
                      "pop": [Function],
                      "push": [Function],
                      "remove": [Function],
                      "removeBatch": [Function],
                      "shift": [Function],
                      "swap": [Function],
                      "unshift": [Function],
                      "update": [Function],
                    },
                    "pauseValidation": [Function],
                    "registerField": [Function],
                    "reset": [Function],
                    "resetFieldState": [Function],
                    "resumeValidation": [Function],
                    "setConfig": [Function],
                    "submit": [Function],
                    "subscribe": [Function],
                  }
                }
              >
                <Component>
                  <div>
                    <Component
                      key="form-submit"
                      label="Submit"
                      onClick={[Function]}
                      type="button"
                      variant="primary"
                    >
                      <button
                        onClick={[Function]}
                        type="button"
                        variant="primary"
                      >
                        Submit
                      </button>
                    </Component>
                  </div>
                </Component>
              </FormSpy>
            </ReactFinalForm(FormSpy)>
          </FormControls>
        </form>
      </Component>
    </ReactFinalForm>
  </FormRenderer>
</ContextWrapper>
`;

exports[`<FormControls /> should render without title and description 1`] = `
<ContextWrapper>
  <FormRenderer
    buttonsLabels={Object {}}
    clearOnUnmount={false}
    disableSubmit={Array []}
    formFieldsMapper={Object {}}
    formType="pf3"
    initialValues={Object {}}
    layoutMapper={
      Object {
        "Button": [Function],
        "ButtonGroup": [Function],
        "Col": [Function],
        "Description": [Function],
        "FormGroup": [Function],
        "FormWrapper": [Function],
        "Title": [Function],
      }
    }
    onSubmit={[MockFunction]}
    resetAble={false}
    schema={
      Object {
        "fields": Array [],
      }
    }
    schemaType="default"
    showFormControls={true}
    uiSchema={Object {}}
  >
    <ReactFinalForm
      decorators={
        Array [
          [Function],
        ]
      }
      initialValues={Object {}}
      mutators={
        Object {
          "concat": [Function],
          "insert": [Function],
          "move": [Function],
          "pop": [Function],
          "push": [Function],
          "remove": [Function],
          "removeBatch": [Function],
          "shift": [Function],
          "swap": [Function],
          "unshift": [Function],
          "update": [Function],
        }
      }
      onSubmit={[MockFunction]}
      render={[Function]}
      subscription={
        Object {
          "pristine": true,
          "submitting": true,
          "valid": true,
        }
      }
    >
      <Component>
        <form>
          <FormControls
            buttonOrder={
              Array [
                "submit",
                "reset",
                "cancel",
              ]
            }
            canReset={false}
            canSubmit={false}
            cancelLabel="Cancel"
            handleSubmit={[Function]}
            resetLabel="Reset"
            submitLabel="Submit"
          >
            <ReactFinalForm(FormSpy)>
              <FormSpy
                reactFinalForm={
                  Object {
                    "batch": [Function],
                    "blur": [Function],
                    "change": [Function],
                    "destroyOnUnregister": false,
                    "focus": [Function],
                    "getFieldState": [Function],
                    "getRegisteredFields": [Function],
                    "getState": [Function],
                    "initialize": [Function],
                    "isValidationPaused": [Function],
                    "mutators": Object {
                      "concat": [Function],
                      "insert": [Function],
                      "move": [Function],
                      "pop": [Function],
                      "push": [Function],
                      "remove": [Function],
                      "removeBatch": [Function],
                      "shift": [Function],
                      "swap": [Function],
                      "unshift": [Function],
                      "update": [Function],
                    },
                    "pauseValidation": [Function],
                    "registerField": [Function],
                    "reset": [Function],
                    "resetFieldState": [Function],
                    "resumeValidation": [Function],
                    "setConfig": [Function],
                    "submit": [Function],
                    "subscribe": [Function],
                  }
                }
              >
                <Component>
                  <div>
                    <Component
                      key="form-submit"
                      label="Submit"
                      onClick={[Function]}
                      type="button"
                      variant="primary"
                    >
                      <button
                        onClick={[Function]}
                        type="button"
                        variant="primary"
                      >
                        Submit
                      </button>
                    </Component>
                  </div>
                </Component>
              </FormSpy>
            </ReactFinalForm(FormSpy)>
          </FormControls>
        </form>
      </Component>
    </ReactFinalForm>
  </FormRenderer>
</ContextWrapper>
`;
